src/librustc_mir/borrow_check/conflict_errors.rs

```diff
@@ -632,33 +631,37 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 // borrowed place and look for a access to a different field of the same union.
                 let mut current = second_borrowed_place;
                 while let Place::Projection(box Projection { base, elem }) = current {
-                    match elem {
-                        ProjectionElem::Field(field, _) if {
-                            is_union(base) && field != target_field && base == target_base
-                        } => {
-                            let desc_base = self.describe_place(base)
-                                .unwrap_or_else(|| "_".to_owned());
-                            let desc_first = self.describe_place(first_borrowed_place)
-                                .unwrap_or_else(|| "_".to_owned());
-                            let desc_second = self.describe_place(second_borrowed_place)
-                                .unwrap_or_else(|| "_".to_owned());
-
-                            // Also compute the name of the union type, eg. `Foo` so we
-                            // can add a helpful note with it.
-                            let ty = base.ty(self.mir, self.infcx.tcx).ty;
-
-                            return Some((desc_base, desc_first, desc_second, ty.to_string()));
-                        },
-                        _ => current = base,
+                    if let ProjectionElem::Field(field, _) = elem {
+                        if let Some(union_ty) = is_union(base) {
+                            if field != target_field && base == target_base {
+                                let desc_base =
+                                    self.describe_place(base).unwrap_or_else(underscore_str_fn);
+                                let desc_first = self
+                                    .describe_place(first_borrowed_place)
+                                    .unwrap_or_else(underscore_str_fn);
+                                let desc_second = self
```

Comment:
Hmm; you are still repeating `self.describe_place($some_place).unwrap_or_else(underscore_str_fn)` three times; can't you have:
```
let describe_place = |place| self.describe_place(place).unwrap_or_else(|| "_".to_owned());
```
or something like that and then write:
```rust
return Some((
    describe_place(base),
    describe_place(first_borrowed_place),
    describe_place(second_borrowed_place),
    union_ty.to_string(),
));
```

Reply:
Fixed, I wasn't paying a lot of attention to the existing code to be honest :).